### ISSUE TITLE
Insurance drain via equal-notional decoy lock + explicit-candidate forced cascade — 4.13 SOL profit, 82.6% fund drain (escalation of #89)

### ISSUE THREAD

**Issue by vip-ultr** [aeyakovenko/percolator-prog#90]
## Overview

This issue escalates the root cause identified in #89 (`RISK_BUF_CAP = 4` risk buffer overflow) via a distinct and more damaging attack path. Where #89 relies on passive buffer overflow through volume (open K > 2 pairs, run empty cranks, wait for Phase C to discover victims slowly), this variant uses two **deliberate equal-notional decoy positions** to permanently lock all four buffer slots, then forces victim liquidations at attacker-chosen timing using the `KeeperCrank` explicit-candidate mechanism. The result is 82.6% insurance fund drain and **4.133 SOL net profit** from a 2.729 SOL investment — compared to 54.5% drain and 2.729 SOL profit in the #89 K=64 case.

The exploit is confirmed against the deployed binary (`6e2bb5aee602aed1de0b2d80f72f97b6b115e0f536438f76d31e0de06d5b7002`, unchanged since #89 was filed) via LiteSVM.

---

## Relationship to Issue #89

Issue #89 identified `RISK_BUF_CAP = 4` as the root cause: with only 4 buffer slots, opening more than 2 leveraged pairs causes the excess pairs to be excluded from the risk buffer and discovered slowly by Phase C (4 accounts per crank). This creates a cascade where later-discovered accounts accumulate multiple segments of adverse price movement before liquidation, generating insurance-covered deficits that grow with each crank.

**The fix suggested in #89** — removing or re-gating the `!candidates.is_empty()` check so Phase-2 RR scan runs unconditionally — would address the *passive* overflow by allowing the engine to discover more at-risk accounts per empty crank.

**This variant bypasses that fix entirely.** It does not rely on empty cranks or Phase C discovery. Instead:

1. The attacker uses the **equal-notional displacement rule** in `buf.upsert` (line ~395) to permanently fill all 4 buffer slots with decoy positions that the engine cannot liquidate (they are above maintenance margin throughout).
2. The attacker then supplies **victim accounts as explicit `FullClose` candidates** in `KeeperCrank`. Explicit candidates are added to the `combined` liquidation list directly (lines 7142–7160), bypassing the risk buffer entirely.
3. By stalling N empty cranks first, the attacker chooses exactly how many segments of adverse price the victims accumulate before being liquidated — making deficits arbitrarily large.

The fix for #89 leaves the equal-notional locking and explicit-candidate liquidation paths completely unaddressed.

---

## Root Cause

Three mechanisms compose:

### 1. `RISK_BUF_CAP = 4` with strict-greater-than displacement

```rust
// percolator-prog/src/percolator.rs  line 32
pub const RISK_BUF_CAP: usize = 4;

// line ~395 — upsert eviction logic
if notional <= self.min_notional {
    return false;  // equal-notional entry cannot displace an existing slot
}
```

When the buffer is full and a new account has notional **equal to** (not strictly greater than) the minimum existing entry, the upsert silently fails. The new account is permanently excluded from the buffer for as long as equal-notional accounts occupy all four slots.

### 2. Decoy positions produce equal notional to victim positions

`risk_notional_ceil(eff, crank_price)` is computed from `effective_pos_q` and the current oracle price. Two accounts with identical `pos_q` (position size) will always have identical notional at every crank price, regardless of how much capital (margin) each account carries. An attacker can open positions with the **same `pos_q` as victims** but with **high capital**, making the decoys immune to liquidation (well above maintenance margin) while producing exactly equal notional — permanently blocking victim upserts.

### 3. Explicit `FullClose` candidates bypass the risk buffer

```rust
// percolator-prog/src/percolator.rs  lines 7142–7160
for &(cidx, policy) in candidates.iter() {
    let mut already_promoted = candidate_list_contains(&combined, cidx);
    // ...
    if !already_promoted {
        combined.push((cidx, policy));   // added to combined regardless of buffer membership
    }
    if combined.len() == COMBINED_CAP { break; }
}
```

Caller-supplied candidates are appended to the `combined` liquidation list whether or not they appear in the risk buffer. `MAX_KEEPER_CANDIDATES = 8` allows up to 8 indices per crank. Each is evaluated by the engine: if below maintenance margin, it is liquidated with the full accumulated deficit charged to insurance.

---

## Attack Steps

**Prerequisites:** ~2.73 SOL, any wallet. No privileged access. All instructions permissionless.

### Step 1 — Open 2 decoy pairs (fill the buffer permanently)

Register 2 user accounts and 2 LP accounts. Deposit `200,000,000 lamports` capital per account (generous margin, roughly 0.2 SOL). Open each pair at `pos_q = victim_size` (the same position size the victims will use). These 4 accounts fill all 4 buffer slots on their respective `TradeNoCpi` calls.

Decoy effective leverage ≈ 0.19×. They will not fall below maintenance margin even after 18 adverse price segments (loss = 18 × ~9.3M = ~167M lamports, well below their 200M capital margin).

### Step 2 — Open 60 victim pairs (permanently excluded from buffer)

Register 60 user accounts and 60 LP accounts. Deposit `10,000,000 lamports` per account (0.01 SOL). Open each pair at the same `pos_q = victim_size`. Because all four buffer slots are already occupied by decoys with equal notional, every `buf.upsert` call for a victim silently returns `false`. None of the 120 victim accounts ever enter the risk buffer.

### Step 3 — Stall the oracle (200 slots, 2× adverse price)

Do not call `KeeperCrank` for 200 slots while the oracle moves adversely (for an inverted market: raw oracle increases from ~138M to ~276M baseline). All victim LONG positions are now deeply below maintenance margin. `engine.last_market_slot` is 200 slots behind `clock.slot`.

### Step 4 — Run 10 empty `KeeperCrank` calls (accumulate N=10 oracle segments)

Call `KeeperCrank` with an empty candidate list 10 times. Because the buffer holds the 4 decoy entries (`has_liquidation_hint = true`), each crank advances the oracle by exactly one segment (`max_accrual_dt = 10` slots). Victims are not in the buffer and are not touched. After 10 empty cranks, victims have accumulated 10 full segments of adverse oracle movement without a single liquidation.

### Step 5 — Run 8 candidate `KeeperCrank` calls (liquidate all 60 victims)

Call `KeeperCrank` 8 times, supplying 8 victim user indices as explicit `FullClose` candidates each time (7 batches of 8 + 1 batch of 4 = 60 total). The engine adds each candidate to `combined` (lines 7142–7160). Decoys in the buffer are processed first but not liquidated (above maintenance). Each victim is liquidated at the current oracle price.

At crank N after step 4, each victim has experienced (10 + N) full segments of adverse movement. At 19× leverage:

| Candidate crank | Segment k | Loss per victim | Deficit per victim (capital = 10M) | 8-victim drain |
|---|---|---|---|---|
| 0 | 10–11 | ~931%–1024% capital | ~83.1M–102.4M lamports | ~565M lamports |
| 1 | 11 | ~1024% capital | ~102.4M lamports | ~761M lamports |
| 2 | 12 | ~1117% capital | ~111.7M lamports | ~811M lamports |
| 3 | 13 | ~1210% capital | ~121.0M lamports | ~851M lamports |
| 4 | 14 | ~1303% capital | ~130.3M lamports | ~851M lamports |
| 5 | 15 | ~1396% capital | ~139.6M lamports | ~511M lamports |
| 6 | 16 | ~1489% capital | ~148.9M lamports | ~341M lamports |
| 7 | 17 | ~1582% capital | ~158.2M lamports | ~170M lamports |

### Step 6 — Withdraw LP equity

All 60 LP (short) accounts are profitable. Attacker withdraws all LP equity.

---

## Confirmed Economics

| Item | Lamports | SOL |
|---|---|---|
| Decoy cost: 4 accounts × (200M + 5.882M) | 823,528,000 | 0.824 |
| Victim cost: 120 accounts × (10M + 5.882M) | 1,905,840,000 | 1.906 |
| **Total gross cost** | **2,729,368,000** | **2.729** |
| Insurance drained (`ins_pre` → `ins_final`) | 4,862,154,446 | 4.862 |
| LP equity recovered (conservation: total_slab_start − ins_final) | 6,862,154,446 | 6.862 |
| **Attacker net profit** (LP equity − gross cost) | **4,132,786,446** | **4.133** |
| Loss to original 5 SOL insurance seed | 4,130,427,842 | **4.130 (82.6%)** |

---

## Confirmed Per-Crank Trace (LiteSVM, deployed binary)

```
phase    crank  lms   ins_delta      victims_open  decoy_pos
──────────────────────────────────────────────────────────────────────
empty      0    110          0            60       [26249835312, 26249835312]
empty      1    120          0            60       [unchanged]
empty      2    130          0            60       [unchanged]
empty      3    140          0            60       [unchanged]
empty      4    150          0            60       [unchanged]
empty      5    160          0            60       [unchanged]
empty      6    170          0            60       [unchanged]
empty      7    180          0            60       [unchanged]
empty      8    190          0            60       [unchanged]
empty      9    200          0            60       [unchanged]
──────────────────────────────────────────────────────────────────────
cand       0    210   565,594,800        52       [unchanged]
cand       1    220   760,554,110        42       [unchanged]
cand       2    230   811,489,210        32       [unchanged]
cand       3    240   851,398,870        22       [unchanged]
cand       4    250   851,408,380        12       [unchanged]
cand       5    260   510,850,734         6       [unchanged]
cand       6    270   340,570,960         2       [unchanged]
cand       7    280   170,287,382         0       [unchanged]
──────────────────────────────────────────────────────────────────────
TOTAL insurance drain:  4,862,154,446 lamports (4.862 SOL)
Net profit to attacker: 4,132,786,446 lamports (4.133 SOL)
Victims remaining:      0 / 60
Decoy positions:        open throughout (26,249,835,312 — unchanged)
```

Note on crank[1] liquidating 10 victims instead of 8: after crank[0], Phase C finds 2 newly-freed buffer slots (the two buffer-resident victims that were liquidated) and fills them from the scan window. These 2 buffer-resident victims are deduplicated against the candidate list in the next crank but still appear in `combined`, yielding 2 (from buffer) + 8 (candidates, deduplicated) = 10 unique liquidations from crank[1] onward.

---

## Why the #89 Suggested Fix Does Not Mitigate This

Issue #89 suggests removing the `!candidates.is_empty()` guard so Phase-2 RR scan promotes at-risk accounts even on empty-candidate cranks. This would collapse the passive-overflow cascade by discovering more victims per crank.

This variant is unaffected because:

1. **Explicit candidates are supplied** — the `!candidates.is_empty()` branch is always taken. Phase-2 RR scan already runs.
2. **Buffer is permanently locked** — even if Phase-2 RR scan runs and finds victims, it calls `buf.upsert(victim, notional)` which fails (equal notional to decoys). Victims can never enter the buffer or the Phase-2 RR scan output.
3. **Victims enter `combined` via lines 7142–7160** directly, not via the risk buffer or Phase-2 RR scan.

The only fix that addresses both variants is preventing the buffer from being permanently locked in the first place.

---

## Suggested Fix

The fundamental invariant that must hold: **every open leveraged position must always be trackable by the risk buffer**. Two complementary fixes:

**Option A — Reject trades that would overflow the buffer:** In the `TradeNoCpi` handler, before calling `buf.upsert`, check whether the upsert would succeed. If the buffer is full and the new account's notional does not strictly exceed the minimum existing entry, reject the trade with an error. This prevents the overflow condition from being created.

**Option B — Replace strict-greater-than with greater-than-or-equal eviction:** Change `if notional <= self.min_notional` to `if notional < self.min_notional`. This allows equal-notional accounts to displace existing entries, breaking the permanent lock. Combined with a round-robin or age-based eviction policy, this prevents any single cohort from monopolizing the buffer.

**Option C — Increase `RISK_BUF_CAP`** to at least `MAX_ACCOUNTS / 2` (or the market's configured account limit). This eliminates overflow entirely but significantly increases per-crank state size.

---

## Reproduction

All tests in `percolator-prog-deployed/tests/test_bounty3_sweeps.rs`, compiled against the deployed source (`04b854e`, engine `5059332f`).

```bash
# Primary PoC — decoy forced cascade
cargo test decoy_forced_cascade -- --nocapture

# Standard overflow from issue #89 (for reference)
cargo test sweep_k_scaling -- --nocapture
```

Expected output for `decoy_forced_cascade`:

```
net_profit        = 4132786446 (4.1328 SOL)
ins_fund_loss     = 4130427842 (4.1304 SOL)
VERDICT: PARTIAL DRAIN — 82.61% of Toly's fund drained.
test decoy_forced_cascade ... ok
```

---

## Binary Attestation

| Field | Value |
|---|---|
| Program | `2LfCFmDKwcnHunqdsCW9uV7KNgBgnFGASs8uM7MwHgHm` |
| Slab | `zExGagF9FeMTYGjvkBhknmNzLAP7toX6Aj6Pu1kuvmT` |
| On-chain ELF SHA-256 | `6e2bb5aee602aed1de0b2d80f72f97b6b115e0f536438f76d31e0de06d5b7002` |
| Deployed commit | `04b854e` (percolator-prog), engine `5059332f` |
| Status | Unpatched — binary unchanged since #89 filed 2026-05-10 |

**Comment by aeyakovenko** [comment]
Clean-room addressed on main in ee19b9f.

Accepted/fixed the narrow low-CU cache-quality issue: a full risk buffer previously rejected a new entry whose notional equaled the current minimum, so the first accounts in an equal-notional tie class could pin the secondary cache. The fix changes that tie-break so equal-min entries may replace one existing min slot; strictly lower-notional entries are still rejected.

Regression added: test_buffer_equal_notional_tie_refreshes_cache. It proves a full equal-notional buffer admits a new equal-notional account, keeps count/min stable, and replaces exactly one old tie entry. I also updated the existing eviction test to cover the equal-min refresh behavior.

Verification run locally:
- cargo test --release --test test_risk_buffer equal_notional -- --nocapture (red before fix, green after)
- cargo build-sbf --no-default-features
- cargo test --release --test test_risk_buffer -- --nocapture
- cargo test --release --test test_security issue65 -- --nocapture
- cargo test --release --test test_security issue81 -- --nocapture

I am not accepting the posted insurance-drain economics as validated against current main. The risk buffer is a secondary discovery cache; primary crank/liquidation safety remains engine/keeper-owned and is covered by the issue65/issue81 regressions above. Closing this as addressed for the valid equal-notional cache tie-break.
